### PR TITLE
sample sheet - override cycles, barcode mismatches and hamming distance fixes

### PR DESCRIPTION
**Pull request by diitaz93**
## Description

This PR addresses three things described in #2449 :
1. Fix the barcode mismatch assigning for NovaSeq6000, as it was set to 1 (default) for all cases. The problem was that the correct barcode mismatch assigning was only done for NovaSeqX.
2. Implement the assigning of OverrideCycles for BCLConvert sample sheets.
3. Fix the hamming distance calculation based on the discussion in https://github.com/Clinical-Genomics/cg_lims/issues/392

### Added

- Functions to calculate the hamming distance independently for index1 (i7) and index2 (i5) according to what we agreed upon on https://github.com/Clinical-Genomics/cg_lims/issues/392
- Tests for the new hamming distance calculators
- Functions to update barcode mismatch values for each sample in each SampleSheetCreator class
- Functions to calculate override cycles for each sample in each SampleSheetCreator class
- Attribute `OverrideCycles` to FlowCellSampleBCLConvert` model

### Changed

- Splitted the function `adapt_samples` into `update_indexes_for_samples` and `update_barcode_mismatch_values_for_samples`
- Renamed function `adapt_indexes_for_sample` to `pad_and_reverse_complement_sample_indexes`

### Fixed

- Removed if statement that excluded NovaSeq6000 barcode mismatches to be updated
- Modified the checks for padding to only pass when the samples are from a Bcl2fastq flow cell
- Removed repeated fixtures and updated usages

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b sample-sheet-override-and-barcode -a
    ```

### How to test

- [x] Delete a sample sheet from a flow cell on stage and from HK
- [x] `cg -l DEBUG demultiplex samplesheet create <flow-cell-name> --dry-run``
- [x] Do this for BCLConvert and Bcl2fastq

### Expected test outcome

- [x] Verify that the sample sheet was created correctly
- [x] Take a screenshot and attach or copy/paste the output.

## Review

- [x] Tests executed by SD
- [x] "Merge and deploy" approved by VJ
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan

- [ ] Deploy this branch on 
  - [ ] stage
  - [ ] production

### CODE REVIEW

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (ChrOertlin) [inline cg/apps/demultiplex/sample_sheet/index.py:153]:
```suggestion
    limit: int = get_sequence_limit(sequence_1=sequence_1, sequence_2=sequence_2)
  
def get_sequence_limit(sequence_1: int, sequence_2: int) -> int:
    return min(len(sequence_1), len(sequence_2))
```

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (ChrOertlin) [inline cg/apps/demultiplex/sample_sheet/index.py:161]:
```suggestion
    limit: int = [min(len(sequence_1), len(sequence_2))](limit: int = get_sequence_limit(sequence_1=sequence_1, sequence_2=sequence_2))
```

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (ChrOertlin) [reply cg/apps/demultiplex/sample_sheet/index.py:153]: With next comment - keep it dry

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (ChrOertlin) [inline cg/apps/demultiplex/sample_sheet/index.py:162]: Not sure if this is needed to know but we could clarify why we flip and sign the limit here compared to index_1?

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (ChrOertlin) [inline cg/apps/demultiplex/sample_sheet/index.py:205]:
```suggestion
    if isInstance(sample, FlowCellSampleBcl2Fastq) and is_padding_needed(
```
Would this work too? I think `type` fetches the class object

**Review by ChrOertlin — COMMENTED** [review]
Looks good, don't know enough to approve.

**Comment by diitaz93** [comment]
## Tests on stage
Test for NovaSeq6000 BCLConvert
```shell
Sample sheet passed validation
[Header]
FileFormatVersion,2
RunName,H7FYHDRX2
InstrumentPlatform,NovaSeq6000
IndexOrientation,Forward
[Reads]
Read1Cycles,51
Read2Cycles,51
Index1Cycles,8
Index2Cycles,8
[BCLConvert_Settings]
SoftwareVersion,4.1.7
FastqCompressionFormat,gzip
[BCLConvert_Data]
Lane,Sample_ID,Index,Index2,OverrideCycles,AdapterRead1,AdapterRead2,BarcodeMismatchesIndex1,BarcodeMismatchesIndex2
1,ACC10641A24,GATCAACA,TAACATAG,Y51;I8;I8;Y51;,,,1,1
etc.....
```

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:205]: fixed

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:162]: The explanation is quite long to include in the docstring I think. For more info https://github.com/Clinical-Genomics/cg/issues/2449

**Review by Vince-janv — CHANGES_REQUESTED** [review]
Nice work! I might have found a bug though. Please have a look 😊

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:161]: I was thinking of `index1` and `index2` but that would be confusing with the other nomenclature of index1 = i7 and index2 = 15

Comment (ChrOertlin) [reply cg/apps/demultiplex/sample_sheet/index.py:161]: the code: `limit: int = min(len(sequence_1), len(sequence_2))` is repeated below. It can be put in a function which can be called in both places.

File: tests/apps/demultiplex/test_sample_sheet_creator.py
Comment (islean) [inline tests/apps/demultiplex/test_sample_sheet_creator.py:157]: I would ideally like to see a test that checks if the correct cycles are added. This would also catch the potential bug that @Vince-janv have spotted. Might take some effort with the setup though.

File: cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py:252]: Of course, fixed

File: tests/conftest.py
Comment (islean) [inline tests/conftest.py:1391]:
```suggestion
@pytest.fixture
```
for consistency

**Review by islean — COMMENTED** [review]
Looks nice but I think Vincent's comment needs to be addressed as I agree it could be a bug.

File: tests/apps/demultiplex/test_sample_sheet_creator.py
Comment (diitaz93) [reply tests/apps/demultiplex/test_sample_sheet_creator.py:157]: You are totally right. And this has to do with being more strict with tests. I will give it a try

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (islean) [reply cg/apps/demultiplex/sample_sheet/index.py:199]: Hmm good point - not really. I am probably slightly in favour of having the somewhat vaguer "adapt" indexes or something similar and then having smaller methods called "convert_to_reverse_complement" and "pad_indexes" or something like that. It seems clearer to me now than when I first saw it, though. Might have been thrown off by "reverse_complement" being used as a verb (or is it a verb as well?).

Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:199]: yeah, in this context reverse complement is used as a verb. But then yes, is more confusing than I thought 😅  I will consider renaming

File: cg/apps/demultiplex/sample_sheet/models.py
Comment (Karl-Svard) [inline cg/apps/demultiplex/sample_sheet/models.py:52]:
```suggestion
    override_cycles: str = Field("", alias=SampleSheetBCLConvertSections.Data.OVERRIDE_CYCLES.value)
```

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (islean) [reply cg/apps/demultiplex/sample_sheet/index.py:199]: I am okay with you disregarding this, it is only a slight preference and if you prefer this naming, feel free to keep it 😄

File: cg/apps/demultiplex/sample_sheet/models.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/models.py:52]: fixed

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:199]: We can think of a better name later, when we refactor the whole demux code 😊

File: tests/apps/demultiplex/test_sample_sheet_creator.py
Comment (diitaz93) [reply tests/apps/demultiplex/test_sample_sheet_creator.py]: Take a look now @islean

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (karlnyr) [inline cg/apps/demultiplex/sample_sheet/index.py:231]: Is this function not just obsolete? Why not just use pad_and_reverse_complemet_sample_indexes and put the for loop in there?

File: cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py
Comment (karlnyr) [inline cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py:141]: here, I think that semantics is everything. We need to first add the override cycles, then we need to pad_revcomp indexes and after that we need to update the barcode mismatches. Override cycles will tell us how many index cycles the sample has. Padding should then be done afterwards, even though we currently pad based on the sample type. The last thing we want to do is to update the barcode mismatch values, once again, the way all our logic plays together now wont allow us to pad or alter the index length, but if we did, then this would also be a bug. I don't know how comprehensive the test is, so I would simply put it in the semantic order in which it could prevent bugs to begin with.

File: tests/apps/demultiplex/test_index.py
Comment (karlnyr) [inline tests/apps/demultiplex/test_index.py:344]: Any reason to not assert this to be the exact number that you expect it to be? I feel like this is something we cannot accept be ANYTHING but 2. Goes for the tests above as well.

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:231]: for SRP and clarity, I thought the inner function should do the update for one sample and the outer function loop over all functions. This way the functions are easier to test. Do you have any reason why we wanted it as one big function? I think `pad_and_reverse_complement_sample_indexes` is already a bit complex on its own

File: cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py:141]: Agree, fxed

File: tests/apps/demultiplex/test_index.py
Comment (diitaz93) [reply tests/apps/demultiplex/test_index.py:344]: In this case, the distance is actually not 2 but 6. As this is an index2 comparison, it will always compare aligning from the tail. Making the expected distance explicitly equal to 6 in my opinion, is not that meaningful as you as a reviewer would have to manually verify that aligning by the tails gives 6, while the most important thing is to show that the function is not aligning from the head, otherwise it would be zero. I can put it as 6 if you think it would be clearer but honestly, I think the number is not important.

**Comment by diitaz93** [comment]
## New tests on stage deliver correct output for both sample sheets

**Review by Vince-janv — APPROVED** [review]
My comment have been adressed 👍 Please get approval of @karlnyr before merging

File: cg/apps/demultiplex/sample_sheet/index.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:231]: done

File: cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py:252]: agree, fixed

File: tests/apps/demultiplex/test_index.py
Comment (diitaz93) [reply tests/apps/demultiplex/test_index.py:344]: Ok I refactored the test to be clearer

**Comment by diitaz93** [comment]
## Deployed to stage:
```shell
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/js.diazboada
done.
Log deploy... done.
cg, version 49.0.3
θ63° 1d [js.diazboada@hasta:~] [S_base] 3m30s
```
## Deployed to production:
```shell
repository is clean
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/js.diazboada
done.
Log deploy... done.
cg, version 49.0.3
θ63° 1d [js.diazboada@hasta:~] [P_base] 3m30s
```

**Pull request merged**
